src/realm/array_flex.cpp

```diff
@@ -159,29 +159,207 @@ bool ArrayFlex::find_all(const Array& arr, int64_t value, size_t start, size_t e
+        vector &= last_word_mask;
+        if (vector) {
+            int sub_word_index = first_field_marked(width, vector);
+            return start + sub_word_index;
+        }
+    }
+    return end;
+}
+
+// print
+//     bf_iterator it_index{data, static_cast<size_t>(offset), ndx_width, ndx_width, start};
+//     std::cout << "Index: ";
+//     auto start2 = start;
+//     for (; start2 < end; ++start2, ++it_index) {
+//         std::cout << *it_index << ", ";
+//     } std::cout << std::endl;
+//     std::cout << "Value: ";
+//     bf_iterator it_value{data, 0, v_width, v_width, 0};
+//     start2 = start;
+//     for (; start2 < v_size; ++start2, ++it_value) {
+//         std::cout << *it_value << ", ";
+//     }std::cout << std::endl;
+//
+
+bool ArrayFlex::find_eq(const Array& arr, int64_t value, size_t start, size_t end, size_t baseindex,
+                        QueryStateBase* state) const
+{
     const auto& encoder = arr.m_encoder;
+    const auto v_width = encoder.m_v_width;
+    const auto v_size = encoder.m_v_size;
+    const auto ndx_width = encoder.m_ndx_width;
+    // const auto mask = encoder.width_mask();
     const auto data = (uint64_t*)arr.m_data;
+    const auto offset = v_size * v_width;
+
+    auto v_start = parallel_subword_find<Equal>(arr, value, encoder.m_v_mask, 0, v_width, 0, v_size);
+    start = parallel_subword_find<Equal>(arr, v_start, encoder.m_ndx_mask, offset, ndx_width, start, end);
+    bf_iterator it(data, offset, ndx_width, ndx_width, start);
+    for (; start < end; ++start, ++it) {
+        if (*it != v_start)
+            continue;
+        if (!state->match(start + baseindex))
+            return false;
+    }
+
+    // double jump find
+    //    bf_iterator it_index{data, static_cast<size_t>(offset), ndx_width, ndx_width, start};
+    //    for (; start < end; ++start, ++it_index) {
+    //        const auto v = sign_extend_field_by_mask(mask, read_bitfield(data, *it_index * v_width, v_width));
+    //        if (v == value)
+    //            if (!state->match(start + baseindex))
+    //                return false;
+    //    }
+    return true;
+}
+
+bool ArrayFlex::find_neq(const Array& arr, int64_t value, size_t start, size_t end, size_t baseindex,
+                         QueryStateBase* state) const
+{
+    const auto& encoder = arr.m_encoder;
+    const auto v_width = encoder.m_v_width;
+    const auto v_size = encoder.m_v_size;
+    const auto ndx_width = encoder.m_ndx_width;
+    // const auto mask = encoder.width_mask();
+    const auto data = (uint64_t*)arr.m_data;
+    const auto offset = v_size * v_width;
+
+    auto v_start = parallel_subword_find<Equal>(arr, value, encoder.m_v_mask, 0, v_width, 0, v_size);
+    start = parallel_subword_find<NotEqual>(arr, v_start, encoder.m_ndx_mask, offset, ndx_width, start, end);
+    bf_iterator it(data, offset, ndx_width, ndx_width, start);
+    for (; start < end; ++start, ++it) {
```

Comment:
this is weird logic. Using parallel_subword_find to search for the first match, but then a different type of search for subsequent ones? Instead you should have a loop around the last parallel_subword_find